cirq-core/cirq/experiments/readout_confusion_matrix.py

```diff
@@ -0,0 +1,310 @@
+                    qubits in sorted order, i.e. `self.qubits`.
+            method: Correction Method. Should be either 'pseudo_inverse' or 'least_squares'.
+
+        Returns:
+              `(2 ** len(qubits), )` shaped numpy array corresponding to `result` with corrections.
+
+        Raises:
+            ValueError: if `result.shape` != `(2 ** len(qubits),)`.
+        """
+        if qubits is None:
+            qubits = self.qubits
+        if result.shape != (2 ** len(qubits),):
+            raise ValueError(f"result.shape {result.shape} should be {(2 ** len(qubits),)}.")
+        if method not in ['pseudo_inverse', 'least_squares']:
+            raise ValueError(f"method: {method} should be 'pseudo_inverse' or 'least_squares'.")
+
+        if method == 'pseudo_inverse':
+            return result @ self.correction_matrix(qubits)  # coverage: ignore
+
+        # Least squares minimization.
+        cm = self.confusion_matrix(qubits)
+
+        def func(x):
+            print(x.shape)
+            return np.sum((result - x @ cm) ** 2)
+
+        constraints = {'type': 'eq', 'fun': lambda x: sum(result) - sum(x)}
+        bounds = tuple((0, sum(result)) for _ in result)
+        res = scipy.optimize.minimize(
+            func, result, method='SLSQP', constraints=constraints, bounds=bounds
+        )
+        return res.x
+
+    def __repr__(self) -> str:
+        return (
+            f"cirq.ReadoutConfusionMatrix("
+            f"[{','.join([proper_repr(cm) for cm in self.confusion_matrices])}],"
+            f"{self.measure_qubits}"
+            f")"
+        )
+
+    def _json_dict_(self) -> Dict[str, Any]:
+        return {
+            'confusion_matrices': self.confusion_matrices,
+            'measure_qubits': self.measure_qubits,
+        }
+
+    @classmethod
+    def _from_json_dict_(
+        cls, confusion_matrices, measure_qubits, **kwargs
+    ) -> 'ReadoutConfusionMatrix':
+        return cls([np.asarray(cm) for cm in confusion_matrices], measure_qubits)
+
+    def _approx_eq_(self, other: Any, atol: float) -> bool:
+        if not isinstance(other, type(self)):
+            return NotImplemented
+        return self.qubits == other.qubits and all(
+            np.allclose(cm, ocm, atol=atol)
+            for cm, ocm in zip(self.confusion_matrices, other.confusion_matrices)
+        )
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, type(self)):
+            return NotImplemented
+        return self.qubits == other.qubits and all(
+            np.array_equal(cm, ocm)
+            for cm, ocm in zip(self.confusion_matrices, other.confusion_matrices)
+        )
+
+    def __ne__(self, other: Any) -> bool:
+        return not self == other
+
+    def __hash__(self) -> int:
+        vals = tuple(v for cm in self.confusion_matrices for _, v in np.ndenumerate(cm))
+        return hash((ReadoutConfusionMatrix, vals, tuple(self.qubits)))
```

Comment:
this seems expensive. do we need this class to be hashable?

Reply:
I added this to make the `lru_cache` above work. Couldn't find an option to ignore the first parameter in the `lru_cache`, and I think caching the confusion matrices is worth the effort. 

Since the class is immutable (data accesses done via properties), we can probably add a condition here to compute the hash only once and memoize it? Another alternative would be to make the `_confusion_matrix()` an independent method with `lru_cache` and pass the parameters. 

There's also `@methodtools.lru_cache`, but it'll add another dependency to the project. https://pypi.org/project/methodtools/

Reply:
so you definitely don't want to ignore the first parameter for your cached method.

the caching is a performance optimization. instead of re-tensoring the confusion matrix each time that method is called, it can re-use one from before. However, now each call has you tupling and hashing numpy arrays which may be just as expensive!

@dstrain115 pointed out to me that using lru_cache as you have done could lead to a memory leak where the cached matrices survive much longer than the object itself. And the keys of your global cache are tuplified numpy arrays and could use memory as well!

I would probably suggest either not caching that function and waiting for it to become a problem or doing it "the old fashioned way" where you have a private member variable which is a dictionary or something to memoize function calls

Reply:
Removed `lru_cache` and implemented the old fashioned memoization. Also removed `__hash__` for the class.